### PR TITLE
perf: drop the redundant fp64 copies in `clean_data`

### PR DESCRIPTION
**Pull request by arthur-priorlabs**
## Issue

[RES-2240: clean_data: redundant fp64 copies dominate remaining `fit()` CPU preprocessing and drive host-RAM peaks on wide tables](https://linear.app/priorlabs/issue/RES-2240/clean-data-redundant-fp64-copies-dominate-remaining-fit-cpu)

## Motivation and Context

**What**

- `clean_data` no longer makes two redundant full-size float64 copies of the table.
- The predict-time frozen-encoder shortcut is now taken only for a frame that lines up with the fitted one.

**Why**

- Those copies set the RAM peak and most of the remaining CPU: 32 GB transient RSS and 16.8 s on a 5.33 GB float32 table.

**How**

- Drop the unconditional closing `astype(np.float64)`, stop re-copying an already-materialised multi-block frame, and skip the numeric reassign in `fix_dtypes`.
- Gate the shortcut on feature count and feature names, the same two grounds sklearn uses; anything else goes back through `transform`.

______________________________________________________________________

## Public API Changes

- [x] No Public API changes
- [ ] Yes, Public API changes (Details below)

______________________________________________________________________

### Highest supported version

Benchmarked with `scripts/bench_clean_data.py --reference main` from the `arthur/benchmark-clean-data` branch on a `cpuhighmem16spot` node (pandas 3.0.3). The harness also compares outputs against the reference on every run.

> [!WARNING]
> Time deltas are noisy (up to ~7% delta when A/A testing), but RSS isn't (0.5% delta)

Shapes:

- `numeric` is 666,667 x 2,000 float32 (5.33 GB);
- `half-string` and `numeric-object` are 333,333 x 400 object arrays.

"Transient RSS" is peak minus entry, i.e. what the call adds on top of what it was handed; wall time is the median of 3 repeats after 1 warmup.

| mix | transient RSS (main) | transient RSS (this) | Δ RSS | median wall (main) | median wall (this) | Δ time |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| `numeric` | 31.99 GB | 10.67 GB | **−66.7%** | 16.787 s | 5.472 s | **−67.4%** |
| `half-string` | 2.74 GB | 2.74 GB | +0.0% | 14.678 s | 13.990 s | −4.7% |
| `numeric-object` | 3.84 GB | 2.77 GB | **−27.9%** | 10.312 s | 9.418 s | −8.7% |

<details><summary>Benchmark details</summary>

| mix | side | entry RSS | peak RSS | retained | cold call | median | min | max | spread | stdev |
| --- | --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| `numeric` | main | 6.00 GB | 37.99 GB | 10.67 GB | 17.370 s | 16.787 s | 16.598 s | 17.041 s | 2.6% | 222.0 ms |
| `numeric` | this | 6.00 GB | 16.67 GB | 10.67 GB | 5.436 s | 5.472 s | 5.433 s | 5.474 s | 0.7% | 22.9 ms |
| `half-string` | main | 3.88 GB | 6.63 GB | 1.06 GB | 14.595 s | 14.678 s | 14.652 s | 14.800 s | 1.0% | 78.7 ms |
| `half-string` | this | 3.88 GB | 6.63 GB | 1.06 GB | 14.382 s | 13.990 s | 13.806 s | 14.084 s | 2.0% | 141.5 ms |
| `numeric-object` | main | 5.97 GB | 9.82 GB | 1.06 GB | 10.340 s | 10.312 s | 10.291 s | 10.337 s | 0.4% | 22.7 ms |
| `numeric-object` | this | 5.97 GB | 8.74 GB | 1.06 GB | 9.695 s | 9.418 s | 9.411 s | 9.420 s | 0.1% | 5.0 ms |

</details>

Peak RSS on `numeric` falls 37.99 → 16.67 GB (−56.1%), and on `numeric-object` 9.82 → 8.74 GB (−11.0%).

### Lowest supported versions

| mix | transient RSS (main) | transient RSS (this) | Δ RSS | median wall (main) | median wall (this) | Δ time |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| `numeric` (66,667 x 2,000) | 4.31 GB | 1.07 GB | **−75.2%** | 253.827 s | 0.547 s | **−99.8%** |
| `half-string` | 3.36 GB | 3.36 GB | +0.0% | 196.368 s | 197.231 s | +0.4% |
| `numeric-object` | 4.40 GB | 3.32 GB | **−24.5%** | 23.706 s | 22.841 s | −3.6% |

<details><summary>Benchmark explanation</summary>

Re-run against the floors in `pyproject.toml`, resolved with `uv pip install --resolution lowest-direct`: **pandas 1.4.0, numpy 1.21.6, scikit-learn 1.2.0, torch 2.5.0, Python 3.10**. Same harness, same node type, outputs identical to `main` again on all three mixes. `numeric` is measured at 66,667 x 2,000 rather than 666,667 x 2,000 -- at the full shape `main` does not finish one side inside a three-hour allocation, for the reason below.

**Why `numeric` is a different kind of number there.** `fix_dtypes` ends in `X[numerical_columns] = X[numerical_columns].astype(numeric_dtype)`, and pandas < 3 implements a list-key assignment as one single-column `__setitem__` per column (`for k1, k2 in zip(key, value.columns): self[k1] = value[k2]`, `pandas/core/frame.py:3687` in 1.4.0). Each of those does block-manager work proportional to the block count, so the reassign is quadratic in column count.
Timed at 25,000 rows, one call:

| columns | `main` | this branch |
| ---: | ---: | ---: |
| 250 | 0.389 s | 0.022 s |
| 500 | 5.276 s | 0.052 s |
| 1,000 | 28.513 s | 0.143 s |
| 2,000 | 124.470 s | 0.303 s |

Every doubling costs `main` about 4x and this branch about 2x. Skipping the reassign
when it has nothing to do turns it off entirely, which is why the same change is worth
−67% on pandas 3 and −99.8% here.

`half-string` gains nothing on pandas 1.4 (+0.4%) because what it removes is a rounding
error at that scale: the same table takes 196 s there against 14 s on pandas 3, a
difference neither branch controls.

</details>

Added tests:

- `tests/test_preprocessing/test_data_cleaning.py`: the numeric shortcut converts without an intermediate copy, passes non-finites through, and matches the encoder path; `_owned_float64_values` hands a multi-block frame through and still copies a single-block one.
- The frozen-encoder shortcut is pinned three ways -- taken when the frame lines up, rejected for a narrower frame, and reordering like sklearn otherwise -- plus the inf-handling suite over the new numeric path.

______________________________________________________________________

## Checklist

- [x] The changes have been tested locally.
- [x] Documentation has been updated (if the public API or usage changes).
- [x] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
- [x] The code follows the project's style guidelines.
- [x] I have considered the impact of these changes on the public API.

______________________________________________________________________

### CODE REVIEW

**Comment by arthur-priorlabs** [comment]
This change is part of the following stack:

- #1173 ◀
    - #1174

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: src/tabpfn/preprocessing/clean.py [hunk 1]
```diff
@@ -112,6 +143,14 @@ def fix_dtypes(  # noqa: D103, C901, PLR0912
...
             # It's a numeric type, just wrap the array in pandas with the correct dtype
+            #
+            # A wrap, not a copy: when the dtype already matches, the frame returned
+            # from here is backed by the caller's buffer, so it does not own what it
+            # holds. Copy-on-write makes that safe by construction, but pandas 2 is
+            # still supported and there it is not -- a write into such a frame lands
+            # in the caller's array. Every write below either copies the frame first
+            # or replaces whole columns with a new dtype, and `_owned_float64_values`
+            # copies before handing the values back out; keep it that way.
             X = pd.DataFrame(X, copy=False, dtype=numeric_dtype)
```
Comment (jmkuebler) [inline src/tabpfn/preprocessing/clean.py]: I'd tend to remove that comment. IIUC on Pandas 2, this is a reason it is slower. While we want to support older versions, I would say for performance, we do not specifically need to drag their legacy along. (If we deprecate pandas 2 support, this comment will become stale)

File: src/tabpfn/preprocessing/clean.py [hunk 2]
```diff
@@ -173,7 +212,15 @@ def fix_dtypes(  # noqa: D103, C901, PLR0912
...
+    # it rewrites them as one block per column, and a fragmented frame has to be
+    # re-materialised (a full extra copy) by every later `to_numpy`. Skip it when
+    # they already hold the target dtype -- the common case for a numeric ndarray
+    # input, whose DataFrame was constructed with `numeric_dtype` above.
+    if (
+        len(numerical_columns) > 0
+        and not (X.dtypes[numerical_columns] == np.dtype(numeric_dtype)).all()
+    ):
         X[numerical_columns] = X[numerical_columns].astype(numeric_dtype)
     return X
```
Comment (jmkuebler) [inline src/tabpfn/preprocessing/clean.py:218]: Comment is quite verbose. yet, without it, it is very hard to get this in the future. so maybe it is ok to keep. Your choice.

File: src/tabpfn/preprocessing/clean.py [hunk 3]
```diff
@@ -368,6 +415,156 @@ def _inf_masks_dataframe(X: pd.DataFrame) -> tuple[np.ndarray, np.ndarray]:
...
+    through.
+    """
+    if getattr(ord_encoder, "n_features_in_", None) != X.shape[1]:
+        return False
+    fitted_names = getattr(ord_encoder, "feature_names_in_", None)
+    if fitted_names is None or not all(isinstance(col, str) for col in X.columns):
+        return True
+    # Compared as plain lists: the two sides can hold the same names in a numpy
+    # string array and a pandas Index, whose own equality is dtype-sensitive.
+    return list(X.columns) == list(fitted_names)
```
Comment (jmkuebler) [inline src/tabpfn/preprocessing/clean.py]: maybe worthwhile to just do it inline with one or two comments. Otherwise too many lines of code comments for a simple function IMO.

File: src/tabpfn/preprocessing/clean.py (hunk 1 above)
Comment (arthur-priorlabs) [reply src/tabpfn/preprocessing/clean.py]: Maybe now is the time to drop `pandas<3` altogether

**Review by jmkuebler — COMMENTED** [review]
Thanks a lot Arthur! The big gains seem to be achieved by the shorttrack for the cases where there is nothing to encode. This part is also simple to read and maintain. 

Does this already explain the speedups and savings we see for the case of numeric inputs?

On the other inputs the savings are arguably moderate and I am not sure it is then really worth it do introduce that much new code. What's your feeling about this?

If you think it is worthwhile, I'd still aim to get rid of some of the private methods and their verbose docstrings and just do them inline at their unique call sites.

File: src/tabpfn/preprocessing/clean.py (hunk 1 above)
Comment (jmkuebler) [reply src/tabpfn/preprocessing/clean.py]: I think pandas 3 was only released this year, so i guess we will support v2 for a while.

**Comment by arthur-priorlabs** [comment]
Thanks for the review!

The simple fast-track path indeed corresponds to the huge `numeric` (=float-only) gains.

> On the other inputs the savings are arguably moderate

The reason the transient RAM numbers are much lower on `half-string` and `numeric-object` is that the input/output data itself takes much more memory per cell (cf. details section), and so I've had to use much smaller shapes.

I'd argue that -28% RAM usage is pretty significant at the 20B cell scale we're targeting.

Fyi, the mixed dtype paths come into play whenever we have categoricals (either numerical or string), which is pretty common.

File: src/tabpfn/preprocessing/clean.py (hunk 1 above)
Comment (arthur-priorlabs) [reply src/tabpfn/preprocessing/clean.py]: Fixed in b44676ef.

File: src/tabpfn/preprocessing/clean.py (hunk 2 above)
Comment (arthur-priorlabs) [reply src/tabpfn/preprocessing/clean.py:218]: Without it, I'd be the first to always cast everything in the next PR 😅

File: src/tabpfn/preprocessing/clean.py (hunk 3 above)
Comment (arthur-priorlabs) [reply src/tabpfn/preprocessing/clean.py]: Fixed in https://github.com/PriorLabs/TabPFN/pull/1173/commits/e01d537b8be3a811b584b195f38c0f3fbe208f91 + https://github.com/PriorLabs/TabPFN/pull/1173/commits/1d0f80bf5d2ea8a8dbb1dfa92fbb31f50b0fa2e8

**Review by jmkuebler — APPROVED** [review]
Thanks @arthur-priorlabs let's go ahead with this one already (no pandas versioning discussion here IIUC)
